llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td

```diff
@@ -214,76 +214,117 @@ multiclass AMO_C_rr_aq_rl<string clenstr, bits<5> funct5, bits<3> funct3,
-
-let hasSideEffects = 0, mayLoad = 1, mayStore = 1 in
-class CAMO_rr<bits<5> funct5, bit aq, bit rl, bits<3> funct3, string opcodestr>
-    : RVInstRAtomic<funct5, aq, rl, funct3, OPC_AMO,
-                    (outs GPR:$rd), (ins GPCRMemZeroOffset:$rs1, GPR:$rs2),
-                    opcodestr, "$rd, $rs2, $rs1">;
-
-multiclass CAMO_rr_aq_rl<bits<5> funct5, bits<3> funct3, string opcodestr> {
-  def ""     : CAMO_rr<funct5, 0, 0, funct3, opcodestr>;
-  def _AQ    : CAMO_rr<funct5, 1, 0, funct3, opcodestr # ".aq">;
-  def _RL    : CAMO_rr<funct5, 0, 1, funct3, opcodestr # ".rl">;
-  def _AQ_RL : CAMO_rr<funct5, 1, 1, funct3, opcodestr # ".aqrl">;
-}
-
-let hasSideEffects = 0, mayLoad = 1, mayStore = 0 in
-class CLR_C_r<bit aq, bit rl, bits<3> funct3, string opcodestr>
-    : RVInstRAtomic<0b00010, aq, rl, funct3, OPC_AMO,
-                    (outs GPCR:$rd), (ins GPCRMemZeroOffset:$rs1),
-                    opcodestr, "$rd, $rs1"> {
-  let rs2 = 0;
-}
-
-multiclass CLR_C_r_aq_rl<string clenstr, bits<3> funct3, string opcodestr> {
-  def _ # clenstr       : CLR_C_r<0, 0, funct3, opcodestr>;
-  def _AQ_ # clenstr    : CLR_C_r<1, 0, funct3, opcodestr # ".aq">;
-  def _RL_ # clenstr    : CLR_C_r<0, 1, funct3, opcodestr # ".rl">;
-  def _AQ_RL_ # clenstr : CLR_C_r<1, 1, funct3, opcodestr # ".aqrl">;
+multiclass CheriLoad_ri<bits<3> funct3, string opcodestr> {
+  let hasSideEffects = 0, mayLoad = 1, mayStore = 0,
+      DecoderNamespace = "CapModeOnly_" in
+  def "" : RVInstI<funct3, OPC_LOAD, (outs GPR:$rd),
+                   (ins GPCR:$rs1, simm12:$imm12), "c" # opcodestr,
+                   "$rd, ${imm12}(${rs1})">;
+  def : InstAlias<opcodestr # " $rd, ${imm12}(${rs1})",
+                  (!cast<Instruction>(NAME) GPR:$rd, GPCR:$rs1, simm12:$imm12),
+                  0>;
+}
+
+multiclass CheriStore_ri<bits<3> funct3, string opcodestr> {
+  let hasSideEffects = 0, mayLoad = 0, mayStore = 1,
+      DecoderNamespace = "CapModeOnly_" in
+  def "" : RVInstS<funct3, OPC_STORE, (outs),
+                   (ins GPR:$rs2, GPCR:$rs1, simm12:$imm12),
+                   "c" # opcodestr, "$rs2, ${imm12}(${rs1})">;
+  def : InstAlias<opcodestr # "$rs2, ${imm12}(${rs1})",
+                  (!cast<Instruction>(NAME) GPR:$rs2, GPCR:$rs1, simm12:$imm12),
+                  0>;
+}
+
+
+multiclass CLR_r<bit aq, bit rl, bits<3> funct3, string opcodestr,
+                 string Namespace> {
+  let hasSideEffects = 0, mayLoad = 1, mayStore = 0,
+      DecoderNamespace = Namespace in
+  def "" : RVInstRAtomic<0b00010, aq, rl, funct3, OPC_AMO,
+                         (outs GPR:$rd), (ins GPCRMemZeroOffset:$rs1),
+                         "c" # opcodestr, "$rd, $rs1"> {
+    let rs2 = 0;
+  }
+  def : InstAlias<opcodestr # "$rd, $rs1",
+                  (!cast<Instruction>(NAME) GPR:$rd, GPCRMemZeroOffset:$rs1),
+                  0>;
+}
+
+multiclass CLR_r_aq_rl<bits<3> funct3, string opcodestr,
+                       string Namespace = "CapModeOnly_"> {
```

Comment:
Passing the DecoderNamespace is a bit ugly, but I couldn't figure out how to work around InstAlias not having such a field.